File: array/3-Longest_Substring_Without_Repeating_Characters/3_Longest_Substring_Without_Repeating_Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int longestLen = 0;
        int left = 0;
        unordered_map<char, int> charPosMap;
        for (int right = 0; right < s.size(); right++) {
            char c = s[right];
            auto it = charPosMap.find(c);
            left = (it != charPosMap.end() && it->second + 1 > left) ? it->second + 1 : left;
            charPosMap[c] = right;
            longestLen = max(right - left + 1, longestLen);
        }

        return longestLen;
    }
};

File: array/3-Longest_Substring_Without_Repeating_Characters/3_Longest_Substring_Without_Repeating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int longestLen = 0;
        int left = 0;
        Integer prevPos;
        Map<Character, Integer> charPosMap = new HashMap<>();
        for (int right = 0; right < s.length(); right++) {
            Character c = s.charAt(right);
            prevPos = charPosMap.get(c);
            left = (null != prevPos && prevPos+1 > left) ? prevPos+1 : left;
            charPosMap.put(c, right);
            longestLen = Math.max(right - left + 1, longestLen);
        }

        return longestLen;
    }
}
